### ISSUE TITLE
Optionally expose a static Target::builder() -> TargetBuilder method

### ISSUE THREAD

**Issue by TedDriggs** [colin-kiegel/rust-derive-builder#73]
Proposal: On the type where `#[derive(Builder)]` is present, generate an inherent `impl` block for that type which defines a new `builder` method. This has a couple usability perks:

1. The caller doesn't have to import the type and the builder separately.
1. The call is shorter: `RequestOptions::builder()` vs. `RequestOptionsBuilder::default()`

**Comment by colin-kiegel** [comment]
Interesting proposal. :-)
This is closely related to generating a customizable constructor `FooBuilder::new` #69.
cc @faern

Important questions are:
- would the behaviour be the same as in #69?
  - is it opt-in or opt-out?
  - should we offer `skip` / `name` / `private` attributes?
- would it exist in parallel with `FooBuilder::new`? How should we then namespace the settings?
  - `#[builder(constructor(static))]` vs `#[builder(constructor)]`?
    E.g. skipping both `#[builder(constructor(static(skip), skip))]`
  - `#[builder(static_constructor)]` vs `#[builder(constructor)]`?
    E.g. skipping both `#[builder(static_constructor(skip), constructor(skip))]`

A general note on both constructor proposals: If a user of `derive_builder` adds a default value to a previously required field, that would change the signature of every constructor which takes required fields in their signature. I.e. _adding_ default values would be a breaking change for every user of `derive_builder`. We should give a clear hint in the documentation about that and encourage library authors to write short unit tests for their builder API.

**Comment by colin-kiegel** [comment]
My spontaneous tendency would be
- offer both, with namespaced settings like `#[builder(static_constructor)]` and `#[builder(constructor)]`
- make both opt-**out** (i.e. on by default)

**Comment by faern** [comment]
Where does the name `static_constructor` come from? All constructor are static methods on a type. `FooBuilder::default()` is also a static method / constructor.

What need does this `Target::builder()` method fulfill that is currently not supported? To me it sounds like added sugar. Personally I think it's a higher priority to nail the design with infallible build + `TargetBuilder::new` constructor + structured errors + custom validation, which all add important features and might be hard to design nicely.

Con: Occupies a method name on the target type, we otherwise don't touch the target type(?)

**Comment by colin-kiegel** [comment]
![facepalm](https://upload.wikimedia.org/wikipedia/commons/thumb/3/3b/Paris_Tuileries_Garden_Facepalm_statue.jpg/120px-Paris_Tuileries_Garden_Facepalm_statue.jpg)

OMG, you're right about the static-ness of all constructors. Stupid me. ^^'

Yes, it's only sugar for the user not needing to import the builder type. But you could also turn that argument upside-down. E.g. if the user starts with `FooBuilder::new().foo(a).build()?` then he wouldn't need to import the `Foo` type.

Another yes for: we should generally be _very_ careful about what/if we implement anything on the target type. A `builder()` method seems very unlikely to clash with anything in that context of deriving a builder. But generally we have to assume that there can be any method on the target type.

I tend to agree that this would be less important, especially since the argument about not needing to import `FooBuilder` can be fully reversed. Before I realized that, that was the main reason I thought it'd be useful.

**Comment by colin-kiegel** [comment]
I'd like to postpone this idea.

**Comment by TedDriggs** [comment]
> Yes, it's only sugar for the user not needing to import the builder type. But you could also turn that argument upside-down. E.g. if the user starts with `FooBuilder::new().foo(a).build()?` then he wouldn't need to import the `Foo` type.

I'm assuming it's more likely they'll need the base type for a return, a parameter, or a struct declaration. Also, when Rust IDEs start getting struct rename support, it'd be better for the user not to need to remember to find all references to the proc-macro-generated builder.

Would you be opposed to a PR which exposes this as an opt-in struct-level option? It'd be:

`#[builder(inherent_method)]` or `#[builder(inherent_method="ident")]`

There's an early draft (with no attribute) [here](https://github.com/TedDriggs/rust-derive-builder/tree/builder_method).

**Comment by colin-kiegel** [comment]
Maybe later. I think we should add more important things first, like the `new` constructor on `FooBuilder` #69. I also would like to wait a little how the whole API surface grows with the big feature ideas in the pipeline before adding more bells and whistles (and complexity).

Regarding the attribute, `constructor` would be more specific than `method`. I think I'd prefer some `..._constructor`. Not sure what `...` should be though.
